sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,220 @@
-     * @param value Event to publish.
+     * @param message Event to publish.
      */
     @Override
-    protected void hookOnNext(ServiceBusReceivedMessageContext value) {
-        work.next(value);
+    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
+        boolean delivered = false;
+
+        // Boundary condition(timeout case): if timeout has happened, get next available work.
+        if (currentWork == null) {
+            // Now see if there is more queued up work
+            currentWork = getNextWorkAndRequest();
+            logger.verbose("No current work, Picked up next receive request.");
+        }
+
+        if (currentWork != null) {
+            currentWork.next(message);
+            delivered = true;
+            remaining.decrementAndGet();
+            // Check if we have delivered all the messages to current work.
+            if (currentWork.isTerminal()) {
+                currentWork.complete();
+                if (timeoutOperation != null && !timeoutOperation.isDisposed()) {
+                    timeoutOperation.dispose();
+                }
+
+                // Now see if there is more queued up work
+                currentWork = getNextWorkAndRequest();
+                logger.verbose("Current work is terminal, Picked up next receive request.");
+            }
+        }
 
-        if (work.isTerminal()) {
-            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", work.getId());
-            dispose();
+        if (currentWork == null) {
+            if (wip.decrementAndGet() != 0) {
+                logger.warning("There is another worker in drainLoop. But there should only be 1 worker.");
+            }
+            // If message is not delivered to downstream, we will buffer it.
+            if (!delivered) {
+                bufferMessages.add(message);
+            }
         }
     }
 
-    @Override
-    protected void hookOnComplete() {
-        logger.info("[{}] Completed. No events to listen to.", work.getId());
-        dispose();
+    /**
+     *
+     * @param requested credits for current {@link SynchronousReceiveWork}.
+     */
+    private void requestCredits(long requested) {
+        long creditToAdd = requested - (remaining.get() + bufferMessages.size());
+        if (creditToAdd > 0) {
+            remaining.addAndGet(creditToAdd);
+            subscription.request(creditToAdd);
+        }
+    }
+
+    void queueWork(SynchronousReceiveWork work) {
+
+        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", work.getId(), work.getNumberOfEvents(),
+            work.getTimeout());
+        workQueue.add(work);
+        drain();
     }
 
+    private void drain() {
+        if (workQueue.size() == 0) {
+            return;
+        }
+
+        // If someone is already in this loop, then we are already clearing the queue.
+        if (!wip.compareAndSet(0, 1)) {
+            return;
+        }
+
+        // Drain queue..
```

Comment:
The comment I was trying to make about WIP is to use it **only** for if someone is clearing the queue and to control access to the `drainQueue` method. You also modify this value in `hookOnNext`, but never decrement it when we exit the `drainQueue` method.

```java
    private void drain() {
        // If someone is already in this loop, then we are already clearing the queue.
        if (!wip.compareAndSet(0, 1)) {
            return;
        }

        try {
            drainQueue();
        } finally {
            final int decremented = wip.decrementAndGet();
            if (decremented != 0) {
                logger.warning("There should be 0, but was: {}", decremented);
            }
        }
    }
```